core/src/packet.rs

```diff
@@ -510,79 +489,194 @@ impl Blob {
-                let p = r.read().unwrap();
-                let a = p.meta.addr();
-                if let Err(e) = socket.send_to(&p.data[..p.meta.size], &a) {
-                    warn!(
-                        "error sending {} byte packet to {:?}: {:?}",
-                        p.meta.size, a, e
-                    );
-                    Err(e)?;
+
+        self.packets.truncate(i);
+        inc_new_counter_info!("packet-recv_count", i);
+
+        Ok(i)
+    }
+
+    fn send_to(&self, socket: &UdpSocket) -> io::Result<()> {
+        self.packets.send_to(socket)
+    }
+}
+
+impl<M: Message> Mailbox for Vec<M> {
+    type Message = M;
+
+    fn recv_from(&mut self, socket: &UdpSocket) -> io::Result<usize> {
+        let mut i = 0;
+        socket.set_nonblocking(false)?;
+
+        while i < M::BATCH_SIZE {
+            let mut message = M::default();
+
+            match message.recv_from(socket) {
+                Err(_) if i > 0 => {
+                    break;
+                }
+                Err(e) => {
+                    trace!("recv_from err {:?}", e);
+                    return Err(e);
+                }
+                Ok(()) => {
+                    if i == 0 {
+                        socket.set_nonblocking(true)?;
+                    }
+
+                    i += 1;
                 }
             }
+
+            self.push(message);
+        }
+        self.truncate(i);
+        Ok(i)
+    }
+
+    fn send_to(&self, socket: &UdpSocket) -> io::Result<()> {
+        for message in self {
+            //let p = shared_message.write().unwrap();
+
+            //let a = p.meta().addr();
+            //socket.send_to(&p.content()[..p.meta().size], &a)?;
+            message.send_to(socket)?;
         }
         Ok(())
     }
 }
 
-pub fn index_blobs(blobs: &[SharedBlob], id: &Pubkey, mut blob_index: u64, slot: u64, parent: u64) {
-    // enumerate all the blobs, those are the indices
-    for blob in blobs.iter() {
-        let mut blob = blob.write().unwrap();
+impl<M: Message> Message for Shared<M> {
+    const BATCH_SIZE: usize = M::BATCH_SIZE;
 
-        blob.set_index(blob_index);
-        blob.set_slot(slot);
-        blob.set_parent(parent);
-        blob.set_id(id);
-        blob_index += 1;
+    fn recv_from(&mut self, socket: &UdpSocket) -> io::Result<()> {
+        self.write().unwrap().recv_from(socket)
+    }
+
+    fn send_to(&self, socket: &UdpSocket) -> io::Result<()> {
+        self.read().unwrap().send_to(socket)
+    }
+}
+
+impl Message for Packet {
+    const BATCH_SIZE: usize = 1024;
+
+    fn recv_from(&mut self, socket: &UdpSocket) -> io::Result<()> {
+        let (nrecv, from) = socket.recv_from(&mut self.data)?;
+        self.meta.size = nrecv;
+        self.meta.set_addr(&from);
+        Ok(())
+    }
+
+    fn send_to(&self, socket: &UdpSocket) -> io::Result<()> {
+        socket.send_to(&self.data[..self.meta.size], self.meta.addr())?;
+        Ok(())
+    }
+}
+
+impl Message for Blob {
```

Comment:
Why continue to have two types `Packet` and `Blob`?  If `BATCH_SIZE` is the only difference, could `Message` be a struct with two different constructors?